File: clients/go/operator/operator_test.go
package operator

import (
	"bufio"
	"context"
	"encoding/json"
	"io"
	"net"
	"os"
	"path/filepath"
	"testing"
)

const operatorTestCampaign = "018f2f85-3c20-7a31-8f11-112233445566"
const operatorTestSource = "018f2f85-3c20-7a31-8f11-112233445577"
const operatorTestTarget = "018f2f85-3c20-7a31-8f11-112233445588"
const operatorTestHash = "0101010101010101010101010101010101010101010101010101010101010101"

func TestOperatorSessionIsCampaignBoundAndDisjointFromApplicationCalls(t *testing.T) {
	base := os.Getenv("TMPDIR")
	if base == "" {
		t.Fatal("TMPDIR must name the protected test scratch root")
	}
	directory, err := os.MkdirTemp(base, "riffdb-go-operator-")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(directory)
	path := filepath.Join(directory, "operator.sock")
	listener, err := net.Listen("unix", path)
	if err != nil {
		t.Fatal(err)
	}
	defer listener.Close()

	serverDone := make(chan error, 1)
	go func() {
		connection, acceptErr := listener.Accept()
		if acceptErr != nil {
			serverDone <- acceptErr
			return
		}
		defer connection.Close()
		reader := bufio.NewReader(connection)
		for index := 0; index < 3; index++ {
			body, readErr := readOperatorTestFrame(reader)
			if readErr != nil {
				serverDone <- readErr
				return
			}
			var request map[string]any
			if json.Unmarshal(body, &request) != nil || request["operation"] != nil || request["credential"] != nil {
				serverDone <- io.ErrUnexpectedEOF
				return
			}
			requestID, _ := request["request_id"].(string)
			var response any
			switch index {
			case 0:
				response = operatorHandshakeResponse{Type: "handshake", RequestID: requestID, ProtocolVersion: OperatorProtocolVersion, DriverIdentity: "riffdb-driver-host/v1", Database: "restored", CampaignID: operatorTestCampaign, PortabilityManifestHash: operatorTestHash}
			case 1:
				response = struct {
					Type      string            `json:"type"`
					RequestID string            `json:"request_id"`
					Operation ReimportOperation `json:"operation"`
				}{Type: "operation", RequestID: requestID, Operation: testReimportOperation()}
			default:
				response = struct {
					Type      string `json:"type"`
					RequestID string `json:"request_id"`
				}{Type: "not_found", RequestID: requestID}
			}
			encoded, encodeErr := marshalOperator(response)
			if encodeErr != nil {
				serverDone <- encodeErr
				return
			}
			prefix := []byte{byte(len(encoded) >> 24), byte(len(encoded) >> 16), byte(len(encoded) >> 8), byte(len(encoded))}
			if _, writeErr := connection.Write(append(prefix, encoded...)); writeErr != nil {
				serverDone <- writeErr
				return
			}
		}
		serverDone <- nil
	}()

	identity := OperatorIdentity{Database: "restored", CampaignID: operatorTestCampaign, PortabilityManifestHash: operatorTestHash}
	session, err := ConnectOperator(context.Background(), path, identity)
	if err != nil {
		t.Fatal(err)
	}
	defer session.Close()
	operation, err := session.Start(context.Background(), "{}", "{}", 3)
	if err != nil || operation.RowsApplied != "2" {
		t.Fatalf("unexpected operation: %#v %v", operation, err)
	}
	status, err := session.Status(context.Background())
	if err != nil || status != nil {
		t.Fatalf("unexpected status: %#v %v", status, err)
	}
	if err := <-serverDone; err != nil {
		t.Fatal(err)
	}
}

func TestOperatorPageValidationRejectsCallerSelectedUnsafeShapes(t *testing.T) {
	if validOperatorIdentity(OperatorIdentity{Database: "restored", CampaignID: operatorTestCampaign, PortabilityManifestHash: operatorTestHash}) != true {
		t.Fatal("valid identity rejected")
	}
	if validOperatorIdentity(OperatorIdentity{Database: "restored", CampaignID: operatorTestCampaign, PortabilityManifestHash: "bad"}) {
		t.Fatal("invalid manifest hash accepted")
	}
	page := ReimportPage{ExportOperationID: operatorTestSource, PageNumber: 1, CanonicalJSONLines: []string{"{\"entity\":\"Ticket\"}"}, ClassComplete: true, OperationComplete: true, PageHashHex: operatorTestHash, MaximumAttempts: 3}
	if !validReimportPage(page) {
		t.Fatal("valid page rejected")
	}
	cursor := "cursor"
	page.NextCursorBase64 = &cursor
	if validReimportPage(page) {
		t.Fatal("ambiguous terminal cursor accepted")
	}
}

func testReimportOperation() ReimportOperation {
	return ReimportOperation{
		CampaignID: operatorTestCampaign, ContractLineage: "TicketDesk", Scope: "whole_application",
		PortabilityManifestHash: operatorTestHash, ExportManifestHash: operatorTestHash, ExportReceiptHash: operatorTestHash,
		SourceDatabaseID: operatorTestSource, TargetDatabaseID: operatorTestTarget, SourceRows: "3", SourcePages: "1",
		NextPage: "2", RowsApplied: "2", Phase: "applying", Failure: nil, CanonicalReimportReceiptJSON: nil, ReimportReceiptHash: nil,
	}
}

func readOperatorTestFrame(reader io.Reader) ([]byte, error) {
	prefix := make([]byte, 4)
	if _, err := io.ReadFull(reader, prefix); err != nil {
		return nil, err
	}
	length := int(prefix[0])<<24 | int(prefix[1])<<16 | int(prefix[2])<<8 | int(prefix[3])
	body := make([]byte, length)
	_, err := io.ReadFull(reader, body)
	return body, err
}

File: clients/typescript/runtime/src/operator.test.ts
import assert from "node:assert/strict";
import { mkdtemp, rm } from "node:fs/promises";
import { createServer, type Socket } from "node:net";
import { join } from "node:path";
import test from "node:test";

import { OperatorTransport } from "./operator.js";

const CAMPAIGN = "018f2f85-3c20-7a31-8f11-112233445566";
const SOURCE = "018f2f85-3c20-7a31-8f11-112233445577";
const TARGET = "018f2f85-3c20-7a31-8f11-112233445588";
const HASH = "0101010101010101010101010101010101010101010101010101010101010101";

test("operator transport is campaign-bound and has no application dispatch", async () => {
  const scratch = process.env.TMPDIR;
  assert.ok(scratch !== undefined && scratch.length > 0, "TMPDIR must name the protected test scratch root");
  const directory = await mkdtemp(join(scratch, "riffdb-ts-operator-"));
  const socketPath = join(directory, "operator.sock");
  const seen: Array<Record<string, unknown>> = [];
  let count = 0;
  const server = createServer((socket) => {
    let input = Buffer.alloc(0);
    socket.on("data", (chunk: Buffer) => {
      input = Buffer.concat([input, chunk]);
      while (input.length >= 4) {
        const length = input.readUInt32BE(0);
        if (input.length < length + 4) return;
        const request = JSON.parse(input.subarray(4, length + 4).toString("utf8")) as Record<string, unknown>;
        input = input.subarray(length + 4);
        assert.equal(Object.hasOwn(request, "operation"), false);
        assert.equal(Object.hasOwn(request, "credential"), false);
        assert.equal(Object.hasOwn(request, "endpoint"), false);
        seen.push(request);
        const requestId = String(request.request_id);
        const response = count === 0
          ? {
              type: "handshake", request_id: requestId, protocol_version: 1,
              driver_identity: "riffdb-driver-host/v1", database: "restored",
              campaign_id: CAMPAIGN, portability_manifest_hash: HASH,
            }
          : count === 1
            ? { type: "operation", request_id: requestId, operation: operation() }
            : { type: "not_found", request_id: requestId };
        count += 1;
        socket.write(frame(response));
      }
    });
  });
  await new Promise<void>((resolve, reject) => {
    server.once("error", reject);
    server.listen(socketPath, resolve);
  });
  try {
    const transport = await OperatorTransport.connect(socketPath, {
      database: "restored", campaignId: CAMPAIGN, portabilityManifestHash: HASH,
    });
    const progress = await transport.start("{}", "{}", 3);
    assert.equal(progress.rowsApplied, 2n);
    assert.equal(await transport.status(), undefined);
    await assert.rejects(
      transport.applyPage(newPage({ operationComplete: true, nextCursorBase64: "cursor" })),
      /invalid RiffDB operator page/u,
    );
    assert.deepEqual(seen.map((request) => request.type), ["handshake", "start", "status"]);
    await transport.shutdown();
  } finally {
    await new Promise<void>((resolve) => server.close(() => resolve()));
    await rm(directory, { recursive: true, force: true });
  }
});

function newPage(overrides: Partial<Parameters<OperatorTransport["applyPage"]>[0]> = {}): Parameters<OperatorTransport["applyPage"]>[0] {
  return {
    exportOperationId: SOURCE,
    pageNumber: 1n,
    canonicalJsonLines: ["{\"entity\":\"Ticket\"}"],
    classComplete: true,
    operationComplete: true,
    pageHashHex: HASH,
    ...overrides,
  };
}

function operation(): Record<string, unknown> {
  return {
    campaign_id: CAMPAIGN,
    contract_lineage: "TicketDesk",
    scope: "whole_application",
    portability_manifest_hash: HASH,
    export_manifest_hash: HASH,
    export_receipt_hash: HASH,
    source_database_id: SOURCE,
    target_database_id: TARGET,
    source_rows: "3",
    source_pages: "1",
    next_page: "2",
    rows_applied: "2",
    phase: "applying",
    failure: null,
    canonical_reimport_receipt_json: null,
    reimport_receipt_hash: null,
  };
}

function frame(value: Record<string, unknown>): Buffer {
  const body = Buffer.from(JSON.stringify(value), "utf8");
  const output = Buffer.allocUnsafe(body.length + 4);
  output.writeUInt32BE(body.length, 0);
  body.copy(output, 4);
  return output;
}
